File: code/204.计数质数.java
import java.util.ArrayList;
import java.util.List;
/*
 * @lc app=leetcode.cn id=204 lang=java
 *
 * [204] 计数质数
 */

// @lc code=start
class Solution {
    public int countPrimes(int n) {
        if (n < 3) {
            return 0;
        }
        if (n == 3) {
            return 1;
        }
        List<Integer> all = new ArrayList<>();
        all.add(2);
        for (int i = 3; i < n; i += 2) {
            boolean isOk = true;
            for (int j = 0; j < all.size() && all.get(j) * all.get(j) <= i; j++) {
                if (i % all.get(j) == 0) {
                    isOk = false;
                    break;
                }
            }
            if (isOk) {
                all.add(i);
            }
        }
        return all.size();
    }
}
// @lc code=end

File: code/204.计数质数.cpp
/*
 * @lc app=leetcode.cn id=204 lang=cpp
 *
 * [204] 计数质数
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution {
public:
    int countPrimes(int n) {
        vector<int> prime;
        vector<int> isPrime(n, 1);
        for (size_t i = 2; i < n; i++) {
            if (isPrime[i]) {
                prime.push_back(i);
            }
            for (size_t j = 0; j < prime.size() && i * prime[j] < n; j++) {
                isPrime[i * prime[j]] = 0;
                if (i % prime[j] == 0) {
                    break;
                }
            }
        }
        return prime.size();
    }
};
// @lc code=end

